fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1354,18 +1339,34 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # Average grad by world_size for consistency with PyTorch DDP.
                 param.grad.data.div_(self.gradient_predivide_factor)
 
-            callback_fn = functools.partial(self._post_reduction_hook, param)
             if param._is_sharded:
-                assert param._is_sharded
                 assert self._reducer is not None
-                grad_chunks = chunk_and_pad(param.grad.data, self.world_size)
+                # Save the unsharded grad for reduction. We will asynchronously accumulate the reduced gradient into
+                # param._saved_grad_shard. If this FSDP module was called multiple times it's possible that multiple
+                # gradient reductions will happen in an undefined order. But addition commutes, so this order doesn't
```

Comment:
since multiple gradient reductions are queued into the same stream, iiuc, the order should be deterministic?

Reply:
I initially made weak assumptions about the reducer. But the current implementation preserves order, as you say, even with bucketing. This could be rephrased. The main observation is that this logic does not require deterministic ordering.